Add `datastoreusage` Command

nepet opened the pull request:
This PR aims to add a __RPC command__ to the __datastore-api__ called `datastoreusage` that is somewhat inspired by the unix command `du`.

`datastoreusage` returns an object that contains a `usage` report and `total` that is the size in byte occupied by the data stored in the `datastore`.

`datastoreusage` accepts the following (optional) parameter:

- `key` is used as a starting point to traverse the `datastore` to collect the size of the data that is stored under this key and all children.
- `depth` gives the max depth of children that are returned in the `usage` report from the (optional) key. `depth` does **NOT** affect the total size, only the depth of the details in `usage` (much like `du`).

### Examples (all from the same datastore):
`datastoreusage`
```json
{
  "usage": [
     {
       "key": [
          "level0",
          "level0.1"
        ],
        "size": 32
      },
      {
        "key": [
          "level0",
          "level0.2"
        ],
        "size": 64
      },
      {
        "key": [
          "level1"
       ],
       "size": 128
     },
  ],
  "total": 224
}
```


`datastoreusage "" 1`
```json
{
  "usage": [
      {
        "key": [
          "level1"
       ],
       "size": 128
     },
  ],
  "total": 224
}
```

`datastoreusage "level0"`
```json
{
  "usage": [
     {
       "key": [
          "level0",
          "level0.1"
        ],
        "size": 32
      },
      {
        "key": [
          "level0",
          "level0.2"
        ],
        "size": 64
      },
  ],
  "total": 96
}
```



`datastoreusage "level0" 0`
```json
{
  "usage": [],
  "total": 96
}
```

Thoughts and comments are appreciated!

rustyrussell commented:
OK, I would omit the level option for now: simply give a root value (or []) and return the bytes used.

Note that bytes here is not an accurate reflection of actual db usage, since there is overhead.  However, make sure you correctly account for keys, as well as values!  (People can store information in keys!)

Also, result for command X should usually be an *object* called similar to X.  Perhaps something like:

```
"datastoreusage": {
    "key": [],
    "self_bytes": 0, /* len(key + data) */
    "total_bytes": 2222 /* self_bytes + all descendents' self_bytes */ 
}
```

rustyrussell commented on lightningd/datastore.c:
```diff
@@ -320,6 +320,46 @@ static struct command_result *json_deldatastore(struct command *cmd,
+						 const jsmntok_t *params)
+{
+	struct json_stream *response;
+	const char **key, **k = NULL;
```
This initialization is wrong BTW, it's **key which needs initializing, not **k!

rustyrussell commented on lightningd/datastore.c:
```diff
@@ -343,6 +349,11 @@ static struct command_result *json_datastoreusage(struct command *cmd,
 	     stmt = wallet_datastore_next(cmd, cmd->ld->wallet, stmt, 
 	     				  &k, &data, &gen)) {
+
+		// Break if we got a key that we are not interested in.
```
Just merged API change means you don't need to do this any more: wallet_datastore_next will do it.

nepet replied:
Rebased to include the API change.

nepet commented:
Rebased onto master to include new wallet API and implemented the proposed changes.

Since the datastore-API currently only allows `data` to be stored for leafs, the database does not store any bytes for ancestors (no key, no data). Therefore the proposed `self_bytes` would be either `self_bytes=0` if a key has descendants, or `self_bytes=total_bytes` if the key is a leaf and stores data.
That's why I left the field out for now, but we can easily add it if needed later on.

```bash
"datastoreusage": {
    "key": [],
    "total_bytes": 2222 /* len(key_blob + data) + all descendents' len(key_blob + data) */ 
}
```

nepet commented:
Oh, and I saved the key as a string in the response, but I'm happy to change it back to an array type if that's more convenient.

vincenzopalazzo reviewed (commented):
Look like the CI is complaining about some whitespace

```
ightningd/datastore.c:319:	     stmt = wallet_datastore_next(cmd, key, stmt, 
lightningd/datastore.c:323:		
lightningd/datastore.c:334:	
Extraneous whitespace found
make: *** [Makefile:521: check-whitespace/lightningd/datastore.c] Error 1
make: *** Waiting for unfinished jobs....
Cloning into '/home/runner/work/lightning/lightning/external/libwally-core'...
Cloning into '/home/runner/work/lightning/lightning/external/lnprototest'...
Cloning into '/home/runner/work/lightning/lightning/external/lowdown'...
```

rustyrussell reviewed (commented):
This is great, but it needs:

1. A Changelog-Added entry saying there's a new RPC command.
2. A test :)

rustyrussell commented on doc/schemas/datastoreusage.request.json:
```diff
@@ -0,0 +1,23 @@
+  "type": "object",
+  "additionalProperties": false,
+  "required": [],
+  "added": "v23.08",
```
This is now v23.11....

rustyrussell commented on CHANGELOG.md:
```diff
@@ -3,6 +3,12 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## [Unreleased]
+
+### Added
+
+- JSON-RPC: `datastoreusage`: returns the total bytes that are stored under a given key.
+
 
```
No, don't edit CHANGELOG.md!

rustyrussell replied:
Use a Changelog-Added line in your commit message instead.

nepet replied:
Uh, sorry about that, I added a `Changelog-Added` line to the initial commit message.

rustyrussell commented on doc/schemas/datastoreusage.request.json:
```diff
@@ -0,0 +1,23 @@
+  "type": "object",
+  "additionalProperties": false,
+  "required": [],
+  "added": "v23.11",
+  "properties": {
+    "key": {
+      "oneOf": [
+        {
+          "type": "array",
+          "description": "key is an array of values (though a single value is treated as a one-element array). Used as the staring point to traverse the datastore.",
```
typo nit: "starting point"

rustyrussell commented on lightningd/datastore.c:
```diff
@@ -298,34 +298,40 @@ static struct command_result *json_datastoreusage(struct command *cmd,
-	u64 gen;
-	u64 total = 0;
+	u64 gen, total_bytes = 0;
 
```
Combine first two commits, since the first doesn't compile anywayk, and you radically change behavior.

Also **key doesn't need initialization.

rustyrussell commented on lightningd/datastore.c:
```diff
@@ -298,34 +298,40 @@ static struct command_result *json_datastoreusage(struct command *cmd,
+
+		// We use the formatted datastore key to calculate the size of
+		// the key to account for the delimiter byte in the database.
+		self_bytes += strlen(datastore_key_fmt(tmpctx, k)) - 2;
```
Doing the fmt to get the right answer seems strange here?

How about:

```
u64 self_bytes = tal_bytelen(data);
/* One byte per separator */
self_bytes += tal_count(key) - 1;
for (size_t i = 0; i < tal_count(key); i++) {
        self_bytes += strlen(key[i]);
```

nepet replied:
Agreed, It now is calculated form the key chars itself not the formatted string. This is straight forward and makes it even easier as it safes us a weird `-2` or `-1` to remove separators from the calculation.

rustyrussell commented on lightningd/datastore.c:
```diff
@@ -289,6 +289,51 @@ static struct command_result *json_deldatastore(struct command *cmd,
+
+		u64 self_bytes = tal_bytelen(data);
+		for (size_t i = 0; i < tal_count(k); i++) {
+			self_bytes += strlen(k[i]);
```
This isn't quite right, since the actual space in the db includes the tal_count(k)-1 separators.  Note that tal_count(k) is always >=1 though.

nepet replied:
Thanks for your review!
Uh yeah I guess I was still asleep when I did this yesterday. I forgot to double check the db schema. The PK of the table is the key as a `BLOB` with key entries separated by a string terminator `\0`. I forgot to count these and added a fix `self_bytes += tal_count(k) - 1;`
